tsdb/shard.go

```diff
@@ -1786,154 +1822,255 @@ func (fs *MeasurementFieldSet) IsEmpty() bool {
+		} else {
+			return nil
+		}
 	}
-	err = fs.renameFile(path)
+
+	return fs.renameFile(path)
 }
 
-// marshalMeasurementFieldSet: remove the fields.idx file if no fields
-// otherwise, copy the in-memory version into a protobuf to write to
-// disk
-func (fs *MeasurementFieldSet) marshalMeasurementFieldSet() ([]byte, error) {
-	fs.mu.Lock()
-	defer fs.mu.Unlock()
-	if len(fs.fields) == 0 {
-		// If no fields left, remove the fields index file
-		if err := os.RemoveAll(fs.path); err != nil {
-			return nil, err
+// writeChangesToFile: Write a change file for fields.idx
+func (fscm *measurementFieldSetChangeMgr) writeChangesToFile(first writeRequest) {
+	var err error = nil
+	// Put the errorChannel on which we blocked into a slice to allow more invocations
+	// to share the return code from the file write
+	errorChannels := []errorChannel{first.errorReturn}
+	changes := []FieldChanges{first.changes}
+	// On return, send the error to every go proc that send changes
+	defer func() {
+		for _, c := range errorChannels {
+			c <- err
+			close(c)
+		}
+	}()
+	log, end := logger.NewOperation(fscm.Logger, "failed saving changes", "MeasurementFieldSet")
+	defer end()
+	// Do some blocking IO operations before marshalling the changes,
+	// to allow other changes to be queued up and be captured in one
+	// write operation, in case we are under heavy field creation load
+	fscm.mu.Lock()
+	defer fscm.mu.Unlock()
+	fd, err := os.OpenFile(fscm.Path, os.O_CREATE|os.O_APPEND|os.O_SYNC|os.O_WRONLY, 0666)
+	if err != nil {
+		err = fmt.Errorf("failed opening %s: %w", fscm.Path, err)
+		log.Error("failed saving field index changes", zap.Error(err))
+		return
+	}
+
+	// ensure file closed
+	defer errors2.Capture(&err, func() error {
+		if e := fd.Close(); e != nil {
+			return fmt.Errorf("failed closing %s: %w", fd.Name(), e)
 		} else {
-			return nil, nil
+			return nil
 		}
+	})()
+
+	// Read all the pending field and measurement write or delete
+	// requests
+	for {
+		select {
+		case wr := <-fscm.writeRequests:
+			changes = append(changes, wr.changes)
+			errorChannels = append(errorChannels, wr.errorReturn)
+			continue
+		default:
+		}
+		break
+	}
+	// marshal the alice of slices of field changes
+	var b []byte
+	b, err = marshalFieldChanges(changes...)
+	if err != nil {
+		err = fmt.Errorf("error marshaling changes for %s: %w", fd.Name(), err)
+		log.Error("failed to store field index changes", zap.Error(err))
+		return
+	}
+	if b == nil {
+		// No fields, nothing to do but close the file.
+		return
+	}
+	var fi os.FileInfo
+	if fi, err = fd.Stat(); err != nil {
+		err = fmt.Errorf("unable to get size of %s: %w", fd.Name(), err)
+		return
+	} else if fi.Size() > fscm.goodSize {
+		// If we had a partial write last time, truncate the file to remove it.
+		if err = fd.Truncate(fscm.goodSize); err != nil {
+			err = fmt.Errorf("cannot truncate %s to last known good size of %d after incomplete write: %w", fd.Name(), fscm.goodSize, err)
+			return
+		}
+	}
```

Comment:
The `Truncate` with `O_APPEND` should work as desired on Linux. I'm not entirely sure if it will work properly on Windows. I've found some insinuation that it is problematic, since on Windows Go is really emulating POSIX behavior with the Win32 API, but nothing concrete yet. It would be good to verify this works on Windows. The workaround if it doesn't is to not use `O_APPEND` and do a `fd.Seek(0, os.SEEK_END)` immediately after the `OpenFile` and `Truncate` calls (`Truncate` doesn't update the file pointer).

Reply:
This testing may be a problem; I don't have a Go test environment on a Windows machine.

Reply:
Would we be safer always using the `Seek`?

Reply:
I think the Seek doesn't handle race conditions of other writes, though I don't expect any because of the write requests getting serialized on the channel...